### PR TITLE
DAOS-12612 common: Add properties to control checkpoint behavior

### PR DESCRIPTION
**Pull request by jolivier23**
Adds a controls to select mode and also control frequency and WAL usage threshold.  The latter is presently a noop.

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/control/lib/daos/pool_property.go [hunk 1]
```diff
@@ -340,6 +340,61 @@ func PoolProperties() PoolPropertyMap {
...
+			Property: PoolProperty{
+				Number:      PoolPropertyCheckpointFreq,
+				Description: "WAL Checkpointing frequency",
+				valueHandler: func(s string) (*PoolPropertyValue, error) {
+					rbErr := errors.Errorf("invalid Checkpointing Frequency value %s", s)
+					rsPct, err := strconv.ParseUint(strings.ReplaceAll(s, "%", ""), 10, 64)
+					if err != nil {
+						return nil, rbErr
+					}
+					return &PoolPropertyValue{rsPct}, nil
```
Comment (mjmac) [inline src/control/lib/daos/pool_property.go:364]: Looks like a little copypasta here... ;) This is just a simple integer, right? Should probably remove the stuff having to do with stripping '%', etc.

File: src/control/lib/daos/pool_property.go [hunk 2]
```diff
@@ -340,6 +340,61 @@ func PoolProperties() PoolPropertyMap {
...
+			values: map[string]uint64{
+				"disabled": PoolCheckpointDisabled,
+				"timed":    PoolCheckpointTimed,
+				"lazy":     PoolCheckpointLazy,
+			},
+		},
+		"checkpoint_freq": {
+			Property: PoolProperty{
+				Number:      PoolPropertyCheckpointFreq,
+				Description: "WAL Checkpointing frequency",
```
Comment (mjmac) [inline src/control/lib/daos/pool_property.go:357]: Might want to note what the unit is here, e.g. "WAL Checkpointing frequency, in seconds".

File: src/control/lib/daos/pool_property.go [hunk 3]
```diff
@@ -340,6 +340,61 @@ func PoolProperties() PoolPropertyMap {
...
+				Number:      PoolPropertyCheckpointThresh,
+				Description: "WAL Checkpointing trigger threshold",
+				valueHandler: func(s string) (*PoolPropertyValue, error) {
+					rbErr := errors.Errorf("invalid Checkpointing threshold value %s", s)
+					rsPct, err := strconv.ParseUint(strings.ReplaceAll(s, "%", ""), 10, 64)
+					if err != nil {
+						return nil, rbErr
+					}
+					return &PoolPropertyValue{rsPct}, nil
+				},
```
Comment (mjmac) [inline src/control/lib/daos/pool_property.go:387]: I guess this is actually a percentage? In any case, might want to make the description more descriptive.

File: src/pool/srv_layout.c [hunk 4]
```diff
@@ -43,72 +43,104 @@ RDB_STRING_KEY(ds_pool_prop_, scrub_freq);
...
-	}, {
-		.dpe_type	= DAOS_PROP_PO_EC_CELL_SZ,
-		.dpe_val	= DAOS_EC_CELL_DEF,
-	}, {
-		.dpe_type	= DAOS_PROP_PO_REDUN_FAC,
-		.dpe_val	= DAOS_PROP_PO_REDUN_FAC_DEFAULT,
-	}, {
-		.dpe_type	= DAOS_PROP_PO_EC_PDA,
-		.dpe_val	= DAOS_PROP_PO_EC_PDA_DEFAULT,
+    {
```
Comment (mjmac) [inline src/pool/srv_layout.c:52]: Not sure what happened here... Tabs vs. spaces thing?

Comment (jolivier23) [reply src/pool/srv_layout.c:52]: I am using clang-format with our settings in .clang-format so it modifies stuff you touch.  I can probably revert if preferred but I think we are moving in this direction and this is a reasonable change.

Comment (mjmac) [reply src/pool/srv_layout.c:52]: OK. I was just wondering why so much had changed here. Might be worth having a flag day and just running the formatter over everything in one commit? In any case, I'm a big proponent of auto-formatting.

**Review by gnailzenh — APPROVED** [review]

File: src/pool/srv_layout.h [hunk 5]
```diff
@@ -73,6 +73,9 @@ extern d_iov_t ds_pool_prop_scrub_freq;		/* uint64_t */
 extern d_iov_t ds_pool_prop_scrub_thresh;	/* uint64_t */
 extern d_iov_t ds_pool_prop_svc_redun_fac;	/* uint64_t */
 extern d_iov_t ds_pool_prop_obj_version;	/* uint32_t */
+extern d_iov_t ds_pool_prop_checkpoint_mode;    /* uint64_t */
```
Comment (wangshilong) [inline src/pool/srv_layout.h:76]: [minor] do we really need it uint64? we might have only few modes, maybe uint8 is fine?

**Review by wangshilong — APPROVED** [review]

File: src/pool/srv_layout.h (hunk 5 above)
Comment (jolivier23) [reply src/pool/srv_layout.h:76]: Yeah, you are right.  Most of this is just copy paste from other properties

**Pull request merged**